src/MPS.jl

```diff
@@ -0,0 +1,330 @@
+        χl, χr = let after_mid = i > n ÷ 2, i = (n + 1 - abs(2i - n - 1)) ÷ 2
+            χl = min(χ, p^(i - 1))
+            χr = min(χ, p^i)
+
+            # swap bond dims after mid and handle midpoint for odd-length MPS
+            (isodd(n) && i == n ÷ 2 + 1) ? (χl, χl) : (after_mid ? (χr, χl) : (χl, χr))
+        end
+
+        # orthogonalize by QR factorization
+        F = lq!(rand(rng, T, χl, p * χr))
+
+        reshape(Matrix(F.Q), χl, p, χr)
+    end
+
+    # reshape boundary sites
+    arrays[1] = reshape(arrays[1], p, p)
+    arrays[n] = reshape(arrays[n], p, p)
+
+    return MPS(arrays; order=(:l, :o, :r))
+end
+
+# TODO deprecate contract(; between) and generalize it to AbstractAnsatz
+"""
+    Tenet.contract!(tn::MPS; between=(site1, site2), direction::Symbol = :left, delete_Λ = true)
+
+For a given [`MPS`](@ref) tensor network, contracts the singular values Λ between two sites `site1` and `site2`.
+The `direction` keyword argument specifies the direction of the contraction, and the `delete_Λ` keyword argument
+specifies whether to delete the singular values tensor after the contraction.
+"""
+@kwmethod contract(tn::MPS; between, direction, delete_Λ) = contract!(copy(tn); between, direction, delete_Λ)
+@kwmethod function contract!(tn::MPS; between, direction, delete_Λ)
+    site1, site2 = between
+    Λᵢ = tensors(tn; between)
+    Λᵢ === nothing && return tn
+
+    if direction === :right
+        Γᵢ₊₁ = tensors(tn; at=site2)
+        replace!(tn, Γᵢ₊₁ => contract(Γᵢ₊₁, Λᵢ; dims=()))
+    elseif direction === :left
+        Γᵢ = tensors(tn; at=site1)
+        replace!(tn, Γᵢ => contract(Λᵢ, Γᵢ; dims=()))
+    else
+        throw(ArgumentError("Unknown direction=:$direction"))
+    end
+
+    delete_Λ && delete!(TensorNetwork(tn), Λᵢ)
+
+    return tn
+end
+@kwmethod contract(tn::MPS; between) = contract(tn; between, direction=:left, delete_Λ=true)
+@kwmethod contract!(tn::MPS; between) = contract!(tn; between, direction=:left, delete_Λ=true)
+@kwmethod contract(tn::MPS; between, direction) = contract(tn; between, direction, delete_Λ=true)
+@kwmethod contract!(tn::MPS; between, direction) = contract!(tn; between, direction, delete_Λ=true)
+
+# TODO rename it to `lanes`?
+function sites(ψ::MPS, site::Site; dir)
+    if dir === :left
+        return site == site"1" ? nothing : Site(id(site) - 1)
+    elseif dir === :right
+        return site == Site(nsites(ψ)) ? nothing : Site(id(site) + 1)
+    else
+        throw(ArgumentError("Unknown direction for MPS = :$dir"))
+    end
+end
+
+@kwmethod function inds(ψ::MPS; at, dir)
+    if dir === :left && at == site"1"
+        return nothing
+    elseif dir === :right && at == Site(nlanes(ψ); dual=isdual(at))
+        return nothing
+    elseif dir ∈ (:left, :right)
+        return inds(ψ; bond=(at, sites(ψ, at; dir)))
+    else
+        throw(ArgumentError("Unknown direction for MPS = :$dir"))
+    end
+end
+
+function isisometry(ψ::MPS, site; dir, atol::Real=1e-12)
```

Comment:
""" Checks whether the tensor at `site` of the MPS `psi` is an isometry in the direction `dir = (left|right?)?` (up to tolerance `atol`). 
Eg. for an MPS in right canonical form, isisometry(psi, n; dir=left) should return `true`
"""